### PR TITLE
posix: Implement getline and getdelim

### PR DESCRIPTION
**Pull request by 64**
Fixes managarm/mlibc#18.

### CODE REVIEW

File: options/ansi/generic/stdio-stubs.cpp [hunk 1]
```diff
@@ -1003,46 +1003,69 @@ void perror(const char *string) {
...
+				return nwritten;
+			}
+		}
 
-ssize_t getdelim(char **line, size_t *n, int delim, FILE *file_base) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+		// Double the size of the buffer (but make sure it's at least 1024)
+		capacity = (capacity >= 1024) ? capacity * 2 : 1024;
+		buffer = reinterpret_cast<char *>(getAllocator().reallocate(*line, capacity));
```
Comment (Geertiebear) [inline options/ansi/generic/stdio-stubs.cpp:1056]: In general use ``realloc`` directly, and not frigg's allocator. Only use the latter when actually dealing with something from frigg.

File: options/ansi/generic/stdio-stubs.cpp [hunk 2]
```diff
@@ -1003,46 +1003,69 @@ void perror(const char *string) {
...
 		return -1;
 	}
 
-	char *buffer = nullptr;
-	size_t capacity = 0;
-	if(*line) {
-		buffer = *line;
-		capacity = *n;
-	}
+	constexpr size_t chunk = 1024;
```
Comment (Geertiebear) [inline options/ansi/generic/stdio-stubs.cpp:1019]: This variable seems unused.

File: options/ansi/generic/stdio-stubs.cpp [hunk 3]
```diff
@@ -1003,46 +1003,69 @@ void perror(const char *string) {
...
+	auto c = fgetc_unlocked(stream);
+	if (c == EOF || ferror(stream)) {
 		return -1;
+	} else {
+		file->unget(c);
+	}
+
+	while (true) {
+		// Fill the buffer
+		while (capacity > 0 && nwritten < capacity - 1) {
```
Comment (Geertiebear) [inline options/ansi/generic/stdio-stubs.cpp:1037]: What happens when capacity is not 0, but the buffer is still NULL? Maybe that wouldn't happen but might still be a good case to catch.

File: tests/posix/getdelim.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	}
+	assert(!ferror(fp));
+	free(line);
+	fclose(fp);
+
+	size_t nchars = 10000;
+	fp = fopen(TEST_FILE, "w");
+	assert(fp);
+	for (int i = 0; i < nchars; i++)
+		fputc('a', fp);
```
Comment (Geertiebear) [inline tests/posix/getdelim.c:34]: We could use`` fwrite()`` here

**Review by Geertiebear — CHANGES_REQUESTED** [review]

File: tests/posix/getdelim.c (hunk 4 above)
Comment (Geertiebear) [reply tests/posix/getdelim.c:34]: Actually, no we can't

**Comment by avdgrinten** [comment]
I think we should actually use the allocator directly for mlibc-internal uses. That way, we know that if something faults in realloc, it is called by user code. (realloc is implemented on top of getAllocator(), not the other way around).

File: options/ansi/generic/stdio-stubs.cpp (hunk 1 above)
Comment (avdgrinten) [inline options/ansi/generic/stdio-stubs.cpp]: We could also just use `frg::small_vector` instead of open coding a small vector.

Comment (avdgrinten) [reply options/ansi/generic/stdio-stubs.cpp]: Never mind, that doesn't work because we need to return the pointer.

**Comment by Geertiebear** [comment]
> I think we should actually use the allocator directly for mlibc-internal uses. That way, we know that if something faults in realloc, it is called by user code. (realloc is implemented on top of getAllocator(), not the other way around).

How useful is this really? Considering we have gdb support and we can usually addr2line. Also is it safe to assume that no code will break if you allocate something with ``getAllocator()`` and free it using ``free()``?

**Review by Geertiebear — APPROVED** [review]
Aside from the ``realloc()`` discussion, LGTM now

**Comment by avdgrinten** [comment]
`free()` is implemented as `getAllocator()->free()`, so that is always safe.

**Comment by Geertiebear** [comment]
Ah alright, well in that case I guess using ``getAllocator()`` is a good idea.

**Comment by 64** [comment]
I think this is ready now

**Comment by avdgrinten** [comment]
Thanks for the PR. I'm merging this now since Geert already approved it.

**Pull request merged**